File: backend/services/analytics_service.py
import logging
from typing import List, Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy import text
from sqlalchemy.exc import OperationalError, ProgrammingError

from models.analytics import (
    DashboardSummaryResponse,
    DistrictCrimeMetric,
    MonthlyCrimeTrend,
    TopCrimeCategory,
    OfficerWorkloadMetric,
    CaseRecordResponse,
    FilterOptionsResponse,
)

logger = logging.getLogger("crimeshield.analytics_service")


class CrimeAnalyticsService:
    @staticmethod
    def get_dashboard_summary(db: Session) -> DashboardSummaryResponse:
        """
        Query high-level KPIs from PostgreSQL view vw_dashboard_summary.
        Returns fallback mock metrics if PostgreSQL table/view is not seeded yet.
        """
        try:
            query = text("SELECT * FROM vw_dashboard_summary LIMIT 1;")
            result = db.execute(query).mappings().first()
            if result:
                return DashboardSummaryResponse(**dict(result))
        except (OperationalError, ProgrammingError, Exception) as e:
            logger.warning(f"PostgreSQL query failed for dashboard summary (using fallback): {str(e)}")

        # Fallback executive metrics
        return DashboardSummaryResponse(
            TotalCases=14285,
            TotalVictims=15820,
            TotalAccused=18450,
            TotalArrests=16200,
            TotalChargesheetedCases=13480,
            PendingCases=805,
            ChargesheetRatePct=94.36,
            ArrestRatePct=87.80,
        )

    @staticmethod
    def get_district_metrics(db: Session) -> List[DistrictCrimeMetric]:
        """
        Query crime breakdown and clearance rates across police districts from vw_crime_by_district.
        """
        try:
            query = text("SELECT * FROM vw_crime_by_district ORDER BY TotalCases DESC;")
            results = db.execute(query).mappings().all()
            if results:
                return [DistrictCrimeMetric(**dict(row)) for row in results]
        except Exception as e:
            logger.warning(f"PostgreSQL query failed for district metrics: {str(e)}")

        # Fallback district data
        return [
            DistrictCrimeMetric(
                DistrictID=1, DistrictName="Bengaluru City", TotalCases=4520,
                HeinousCrimes=820, NonHeinousCrimes=3500, SensationalCrimes=200,
                ChargesheetedCases=4290, ClearanceRatePct=94.9
            ),
            DistrictCrimeMetric(
                DistrictID=2, DistrictName="Mysuru City", TotalCases=2150,
                HeinousCrimes=310, NonHeinousCrimes=1780, SensationalCrimes=60,
                ChargesheetedCases=2010, ClearanceRatePct=93.5
            ),
            DistrictCrimeMetric(
                DistrictID=3, DistrictName="Mangaluru City", TotalCases=1840,
                HeinousCrimes=240, NonHeinousCrimes=1550, SensationalCrimes=50,
                ChargesheetedCases=1720, ClearanceRatePct=93.4
            ),
            DistrictCrimeMetric(
                DistrictID=4, DistrictName="Hubballi-Dharwad", TotalCases=1620,
                HeinousCrimes=210, NonHeinousCrimes=1370, SensationalCrimes=40,
                ChargesheetedCases=1510, ClearanceRatePct=93.2
            ),
            DistrictCrimeMetric(
                DistrictID=5, DistrictName="Belagavi", TotalCases=1480,
                HeinousCrimes=190, NonHeinousCrimes=1260, SensationalCrimes=30,
                ChargesheetedCases=1380, ClearanceRatePct=93.2
            ),
        ]

    @staticmethod
    def get_monthly_trends(db: Session, year: int = 2026) -> List[MonthlyCrimeTrend]:
        """
        Query monthly crime volume and chargesheet rates from vw_crime_by_month.
        """
        try:
            query = text("SELECT * FROM vw_crime_by_month WHERE IncidentYear = :year ORDER BY IncidentMonth ASC;")
            results = db.execute(query, {"year": year}).mappings().all()
            if results:
                return [MonthlyCrimeTrend(**dict(row)) for row in results]
        except Exception as e:
            logger.warning(f"PostgreSQL query failed for monthly trends: {str(e)}")

        # Fallback monthly trend
        months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
        return [
            MonthlyCrimeTrend(
                IncidentYear=year, IncidentMonth=i+1, MonthName=m,
                TotalCases=1100 + (i * 45), HeinousCrimes=180 + (i * 8), ChargesheetedCases=1040 + (i * 42)
            )
            for i, m in enumerate(months[:7])
        ]

    @staticmethod
    def get_top_crimes(db: Session, limit: int = 10) -> List[TopCrimeCategory]:
        """
        Query top crime classifications from vw_top_crimes.
        """
        try:
            query = text("SELECT * FROM vw_top_crimes LIMIT :limit;")
            results = db.execute(query, {"limit": limit}).mappings().all()
            if results:
                return [TopCrimeCategory(**dict(row)) for row in results]
        except Exception as e:
            logger.warning(f"PostgreSQL query failed for top crimes: {str(e)}")

        return [
            TopCrimeCategory(CrimeHeadID=101, CrimeHeadName="Cyber & Financial Fraud", GravityName="Non-Heinous Offence", TotalCases=3420, PctOfTotalCrimes=23.9, ChargesheetRatePct=95.2),
            TopCrimeCategory(CrimeHeadID=102, CrimeHeadName="Narcotics & NDPS Act", GravityName="Heinous Offence", TotalCases=2840, PctOfTotalCrimes=19.8, ChargesheetRatePct=96.8),
            TopCrimeCategory(CrimeHeadID=103, CrimeHeadName="Property Burglary & Theft", GravityName="Non-Heinous Offence", TotalCases=2510, PctOfTotalCrimes=17.5, ChargesheetRatePct=91.4),
            TopCrimeCategory(CrimeHeadID=104, CrimeHeadName="Assault & Public Disorder", GravityName="Non-Heinous Offence", TotalCases=1980, PctOfTotalCrimes=13.8, ChargesheetRatePct=94.1),
            TopCrimeCategory(CrimeHeadID=105, CrimeHeadName="Organized Syndicate Extortion", GravityName="Heinous Offence", TotalCases=1420, PctOfTotalCrimes=9.9, ChargesheetRatePct=97.5),
        ]

    @staticmethod
    def get_officer_workload(db: Session, limit: int = 15) -> List[OfficerWorkloadMetric]:
        """
        Query investigating officer workload and case clearance rates from vw_officer_workload.
        """
        try:
            query = text("SELECT * FROM vw_officer_workload ORDER BY AssignedCases DESC LIMIT :limit;")
            results = db.execute(query, {"limit": limit}).mappings().all()
            if results:
                return [OfficerWorkloadMetric(**dict(row)) for row in results]
        except Exception as e:
            logger.warning(f"PostgreSQL query failed for officer workload: {str(e)}")

        return [
            OfficerWorkloadMetric(EmployeeID=501, KGID="KG-88421", OfficerName="Insp. Rajesh Kumar", RankName="Police Inspector", UnitName="Indiranagar PS", DistrictName="Bengaluru City", AssignedCases=48, PendingInvestigation=3, ChargesheetedCases=45, ClearanceRatePct=93.75),
            OfficerWorkloadMetric(EmployeeID=502, KGID="KG-91204", OfficerName="Insp. Sunita Rao", RankName="Police Inspector", UnitName="Koramangala PS", DistrictName="Bengaluru City", AssignedCases=42, PendingInvestigation=2, ChargesheetedCases=40, ClearanceRatePct=95.23),
            OfficerWorkloadMetric(EmployeeID=503, KGID="KG-74312", OfficerName="Sub-Insp. Vikram Patil", RankName="Sub-Inspector", UnitName="Vijayanagar PS", DistrictName="Mysuru City", AssignedCases=39, PendingInvestigation=4, ChargesheetedCases=35, ClearanceRatePct=89.74),
        ]

    @staticmethod
    def get_cases(db: Session, limit: int = 3000) -> List[CaseRecordResponse]:
        """
        Query detailed FIR case records from casemaster joined with district, unit, crimehead, crimesubhead, and employee.
        """
        try:
            query = text("""
                SELECT 
                    cm.casemasterid as id, 
                    cm.crimeno as firNumber, 
                    TO_CHAR(cm.crimeregistereddate, 'YYYY-MM-DD') as date, 
                    COALESCE(d.districtname, 'Unknown District') as district, 
                    COALESCE(u.unitname, 'Unknown PS') as policeStation, 
                    COALESCE(ch.crimegroupname, 'General Crime') as category, 
                    COALESCE(csh.crimeheadname, 'Other') as subcategory, 
                    COALESCE(csm.casestatusname, 'Under Investigation') as status, 
                    CASE WHEN g.lookupvalue LIKE '%Heinous%' THEN 'High' WHEN g.lookupvalue LIKE '%Sensational%' THEN 'Medium' ELSE 'Low' END as severity, 
                    CASE WHEN g.lookupvalue LIKE '%Heinous%' THEN 'High' WHEN g.lookupvalue LIKE '%Sensational%' THEN 'Medium' ELSE 'Low' END as riskLevel, 
                    False as repeatOffender, 
                    COALESCE(v.ageyear, 35) as victimAge, 
                    'Male' as victimGender, 
                    COALESCE(a.ageyear, 28) as accusedAge, 
                    'Male' as accusedGender, 
                    COALESCE(e.firstname, 'Unassigned') as officerName, 
                    15 as durationDays, 
                    True as arrestCompleted, 
                    COALESCE(EXTRACT(HOUR FROM cm.crimeregistereddate), 14) as hourOfDay 
                FROM casemaster cm 
                LEFT JOIN unit u ON cm.policestationid=u.unitid 
                LEFT JOIN district d ON u.districtid=d.districtid 
                LEFT JOIN crimehead ch ON cm.crimemajorheadid=ch.crimeheadid 
                LEFT JOIN crimesubhead csh ON cm.crimeminorheadid=csh.crimesubheadid 
                LEFT JOIN casestatusmaster csm ON cm.casestatusid=csm.casestatusid 
                LEFT JOIN gravityoffence g ON cm.gravityoffenceid=g.gravityoffenceid 
                LEFT JOIN employee e ON cm.policepersonid=e.employeeid 
                LEFT JOIN victim v ON cm.casemasterid=v.casemasterid 
                LEFT JOIN accused a ON cm.casemasterid=a.casemasterid 
                ORDER BY cm.crimeregistereddate DESC 
                LIMIT :limit;
            """)
            results = db.execute(query, {"limit": limit}).mappings().all()
            if results:
                return [CaseRecordResponse(**dict(row)) for row in results]
        except Exception as e:
            logger.warning(f"PostgreSQL query failed for case records: {str(e)}")

        return []

    @staticmethod
    def get_filter_options(db: Session) -> FilterOptionsResponse:
        """
        Retrieve distinct real filter options directly from PostgreSQL database tables.
        """
        try:
            districts = [r[0] for r in db.execute(text("SELECT DISTINCT districtname FROM district WHERE districtname IS NOT NULL ORDER BY districtname")).fetchall()]
            categories = [r[0] for r in db.execute(text("SELECT DISTINCT crimegroupname FROM crimehead WHERE crimegroupname IS NOT NULL ORDER BY crimegroupname")).fetchall()]
            statuses = [r[0] for r in db.execute(text("SELECT DISTINCT casestatusname FROM casestatusmaster WHERE casestatusname IS NOT NULL ORDER BY casestatusname")).fetchall()]
            officers = [r[0] for r in db.execute(text("SELECT DISTINCT firstname FROM employee WHERE firstname IS NOT NULL ORDER BY firstname")).fetchall()]
            
            # Police stations mapped by district
            ps_rows = db.execute(text("SELECT COALESCE(d.districtname, 'Other') as dist, u.unitname FROM unit u LEFT JOIN district d ON u.districtid=d.districtid WHERE u.unitname IS NOT NULL ORDER BY u.unitname")).fetchall()
            police_stations = {}
            for dist, unit in ps_rows:
                police_stations.setdefault(dist, []).append(unit)
                
            # Subcategories mapped by category
            sc_rows = db.execute(text("SELECT COALESCE(ch.crimegroupname, 'Other') as cat, csh.crimeheadname FROM crimesubhead csh LEFT JOIN crimehead ch ON csh.crimeheadid=ch.crimeheadid WHERE csh.crimeheadname IS NOT NULL ORDER BY csh.crimeheadname")).fetchall()
            subcategories = {}
            for cat, subcat in sc_rows:
                subcategories.setdefault(cat, []).append(subcat)
                
            return FilterOptionsResponse(
                districts=districts,
                police_stations=police_stations,
                categories=categories,
                subcategories=subcategories,
                statuses=statuses,
                officers=officers
            )
        except Exception as e:
            logger.warning(f"PostgreSQL query failed for filter options: {str(e)}")
            
        return FilterOptionsResponse()

File: frontend/services/analyticsService.ts
import { AnalyticsCase, mockAnalyticsCases } from "@/constants/mockAnalyticsData";

export interface DashboardSummaryResponse {
  total_cases: number;
  total_victims: number;
  total_accused: number;
  total_arrests: number;
  total_chargesheeted_cases: number;
  pending_cases: number;
  chargesheet_rate_pct: number;
  arrest_rate_pct: number;
}

export interface DistrictCrimeMetric {
  district_id: number;
  district_name: string;
  total_cases: number;
  heinous_crimes: number;
  crimes_against_women: number;
  cyber_crimes: number;
  chargesheeted_cases: number;
  under_investigation_cases: number;
  clearance_rate_pct: number;
}

export interface MonthlyCrimeTrend {
  incident_year: number;
  incident_month: number;
  month_name: string;
  total_cases: number;
  heinous_crimes: number;
  cyber_crimes: number;
}

export interface TopCrimeCategory {
  crime_head_id: number;
  crime_head_name: string;
  minor_crime_head: string;
  total_cases: number;
  pct_of_total_crimes: number;
  crime_rank: number;
}

export interface OfficerWorkloadMetric {
  employee_id: number;
  kgid: string;
  officer_name: string;
  rank_name: string;
  unit_name: string;
  district_name: string;
  assigned_cases: number;
  pending_investigation: number;
  chargesheeted_cases: number;
  arrests_made: number;
}

export interface FilterOptionsResponse {
  districts: string[];
  police_stations: Record<string, string[]>;
  categories: string[];
  subcategories: Record<string, string[]>;
  statuses: string[];
  officers: string[];
}

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

const CASES_CACHE_KEY = "ksp_live_cases_cache_v1";
const SUMMARY_CACHE_KEY = "ksp_live_summary_cache_v1";
const FILTER_CACHE_KEY = "ksp_live_filter_cache_v1";

export const analyticsService = {
  // Deterministic defaults for SSR / initial hydration (prevents Next.js Hydration failed errors)
  getInitialCases(): AnalyticsCase[] {
    return mockAnalyticsCases;
  },

  getInitialSummary(): DashboardSummaryResponse | null {
    return null;
  },

  getInitialFilterOptions(): FilterOptionsResponse | null {
    return null;
  },

  // Synchronous client-side cache readers to load real metrics immediately after hydration
  getCachedCases(): AnalyticsCase[] | null {
    if (typeof window !== "undefined") {
      try {
        const cached = localStorage.getItem(CASES_CACHE_KEY);
        if (cached) {
          const parsed = JSON.parse(cached);
          if (Array.isArray(parsed) && parsed.length > 0) {
            return parsed;
          }
        }
      } catch (e) {
        console.warn("Error reading live cases from localStorage:", e);
      }
    }
    return null;
  },

  getCachedSummary(): DashboardSummaryResponse | null {
    if (typeof window !== "undefined") {
      try {
        const cached = localStorage.getItem(SUMMARY_CACHE_KEY);
        if (cached) {
          const parsed = JSON.parse(cached);
          if (parsed && typeof parsed.total_cases === "number") {
            return parsed;
          }
        }
      } catch (e) {
        console.warn("Error reading live summary from localStorage:", e);
      }
    }
    return null;
  },

  getCachedFilterOptions(): FilterOptionsResponse | null {
    if (typeof window !== "undefined") {
      try {
        const cached = localStorage.getItem(FILTER_CACHE_KEY);
        if (cached) {
          const parsed = JSON.parse(cached);
          if (parsed && Array.isArray(parsed.districts)) {
            return parsed;
          }
        }
      } catch (e) {
        console.warn("Error reading filter options from localStorage:", e);
      }
    }
    return null;
  },

  async getSummary(): Promise<DashboardSummaryResponse | null> {
    try {
      const res = await fetch(`${API_BASE_URL}/api/v1/analytics/summary`, {
        cache: "no-store",
      });
      if (!res.ok) throw new Error("Failed to fetch summary");
      const data = await res.json();
      if (typeof window !== "undefined" && data && typeof data.total_cases === "number") {
        try {
          localStorage.setItem(SUMMARY_CACHE_KEY, JSON.stringify(data));
        } catch (e) {}
      }
      return data;
    } catch (error) {
      console.warn("FastAPI backend offline or error fetching summary:", error);
      return analyticsService.getInitialSummary();
    }
  },

  async getDistricts(): Promise<DistrictCrimeMetric[]> {
    try {
      const res = await fetch(`${API_BASE_URL}/api/v1/analytics/districts`, {
        cache: "no-store",
      });
      if (!res.ok) throw new Error("Failed to fetch district metrics");
      return await res.json();
    } catch (error) {
      console.warn("FastAPI backend offline or error fetching districts:", error);
      return [];
    }
  },

  async getMonthlyTrends(year = 2026): Promise<MonthlyCrimeTrend[]> {
    try {
      const res = await fetch(`${API_BASE_URL}/api/v1/analytics/monthly-trends?year=${year}`, {
        cache: "no-store",
      });
      if (!res.ok) throw new Error("Failed to fetch monthly trends");
      return await res.json();
    } catch (error) {
      console.warn("FastAPI backend offline or error fetching monthly trends:", error);
      return [];
    }
  },

  async getTopCrimes(limit = 10): Promise<TopCrimeCategory[]> {
    try {
      const res = await fetch(`${API_BASE_URL}/api/v1/analytics/top-crimes?limit=${limit}`, {
        cache: "no-store",
      });
      if (!res.ok) throw new Error("Failed to fetch top crimes");
      return await res.json();
    } catch (error) {
      console.warn("FastAPI backend offline or error fetching top crimes:", error);
      return [];
    }
  },

  async getOfficerWorkload(limit = 15): Promise<OfficerWorkloadMetric[]> {
    try {
      const res = await fetch(`${API_BASE_URL}/api/v1/analytics/officer-workload?limit=${limit}`, {
        cache: "no-store",
      });
      if (!res.ok) throw new Error("Failed to fetch officer workload");
      return await res.json();
    } catch (error) {
      console.warn("FastAPI backend offline or error fetching officer workload:", error);
      return [];
    }
  },

  async getCases(limit = 5000): Promise<AnalyticsCase[]> {
    try {
      const res = await fetch(`${API_BASE_URL}/api/v1/analytics/cases?limit=${limit}`, {
        cache: "no-store",
      });
      if (!res.ok) throw new Error("Failed to fetch cases");
      const data = await res.json();
      if (typeof window !== "undefined" && Array.isArray(data) && data.length > 0) {
        try {
          localStorage.setItem(CASES_CACHE_KEY, JSON.stringify(data));
        } catch (e) {}
      }
      return data;
    } catch (error) {
      console.warn("FastAPI backend offline or error fetching cases:", error);
      return analyticsService.getInitialCases();
    }
  },

  async getFilterOptions(): Promise<FilterOptionsResponse | null> {
    try {
      const res = await fetch(`${API_BASE_URL}/api/v1/analytics/filter-options`, {
        cache: "no-store",
      });
      if (!res.ok) throw new Error("Failed to fetch filter options");
      const data = await res.json();
      if (typeof window !== "undefined" && data && Array.isArray(data.districts)) {
        try {
          localStorage.setItem(FILTER_CACHE_KEY, JSON.stringify(data));
        } catch (e) {}
      }
      return data;
    } catch (error) {
      console.warn("FastAPI backend offline or error fetching filter options:", error);
      return analyticsService.getInitialFilterOptions();
    }
  },
};
